### PR TITLE
Gh-2996: Ability to pass config in gremlin queries

### LINKED ISSUE #2996: Add ability to pass extra config in gremlin queries
**Describe the new feature you'd like**
Using the [with configuration step](https://tinkerpop.apache.org/docs/current/reference/#configuration-steps-with), we could pass in extra config to Gaffer in gremlin queries. This should be able to do things such as set the options field, or set `FederatedOperation.graphIds`.

**Why do you want this feature?**
It would allow for specifying subgraphs when querying a FederatedStore.

**Additional context**
I think we will need to enable [`VariableFeatures`](https://tinkerpop.apache.org/docs/current/reference/#_features)

### PR DESCRIPTION
**Pull request by tb06904**
Adds the ability to override a limited set of graph variables when doing a gremlin query. This utilises the `with()` feature of gremlin that will attach a `OptionsStrategy` to the query that can then be parsed for the relevant options passed in. 

To implement this it required creating a custom traversal mechanism for gremlin `GraphStep`s so that we can override and parse the options on the query. Adding this step also allowed for a small optimisation to ensure any `hasLabel` type queries on the original `GraphStep` are parsed to create a representative Gaffer `View`. The intention is that additional custom traversal steps are added in future to optimise the mapping of gremlin to standard Gaffer operations.

As an example this allows the following:
```java
g.with("userId", "user").V()   // userId extracted to be used in the operations spawned from the query
g.with("dataAuths", "write-access,read-access").V()   // user access controls to apply on the user
g.with("operationOptions", ["graphId:graph1", "opt1:val1"]).V()   // operation options (can add a federated graph Id here)
```

# Related issue

- Resolve #2996

### CODE REVIEW

File: library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraph.java [hunk 1]
```diff
@@ -228,16 +230,24 @@ public GafferPopGraph(final Configuration configuration, final Graph graph) {
...
+        setDefaultVariables(variables);
 
         serviceRegistry = new ServiceRegistry();
         serviceRegistry.registerService(new GafferPopNamedOperationServiceFactory(this));
+
+        // Register custom traversals
+        TraversalStrategies.GlobalCache.registerStrategies(
+            this.getClass(),
+            TraversalStrategies.GlobalCache.getStrategies(this.getClass()).addStrategies(
+                GafferPopGraphStepStrategy.instance()));
```
[the remark this answers is not in the document]
Comment (tb06904) [reply library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraph.java]: Tried to simplify as much as possible with an import and breaking it up

File: library/tinkerpop/src/test/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraphVariablesTest.java [hunk 2]
```diff
@@ -38,22 +40,36 @@ void shouldRemoveValueFromVariables() {
...
+        final String testUserId = "testUserId";
+        final String testDataAuths = "auth1,auth2";
+        final List<String> testOpOptions = Arrays.asList("graphId:graph1", "other:other");
+        final GafferPopGraphVariables graphVariables = (GafferPopGraphVariables) graph.variables();
+        // Expected format they are returned as
+        final String[] expectedDataAuths = {"auth1", "auth2"};
+        final Map<String, String> expectedOpOptions = Stream.of(
+                new SimpleEntry<>("graphId", "graph1"),
+                new SimpleEntry<>("other", "other"))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```
[the remark this answers is not in the document]
Comment (tb06904) [reply library/tinkerpop/src/test/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraphVariablesTest.java]: Found `assertJ` has a `entry` method to help with this (in normal Java this is easy way to do single line init of maps)

**Comment by tb06904** [comment]
> Given the ability to set the user provided in this PR, to what extent does this resolve #2991?

Good question, I would say it probably solves it as it lets an alternative user ID and data auths to be used on a per operation basis. With something like gremlinpy its also possible to make the ID and auths standard on every request like:
```python
# Set opts
withopts = g.with("userId", "user).with("dataAuths", "auth1,auth2")
# Run query
withopts.V().to_list()
```

File: library/tinkerpop/src/test/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraphVariablesTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (GCHQDeveloper314) [reply library/tinkerpop/src/test/java/uk/gov/gchq/gaffer/tinkerpop/GafferPopGraphVariablesTest.java]: Nice, I'd hoped assertJ would have a smart way to do this. It's too bad we're not on Java 11 yet, as while this was the best way to do it in Java 8, it's much easier in 11.

**Review by GCHQDeveloper314 — DISMISSED** [review]

**Review by l46978 — APPROVED** [review]

**Review by GCHQDeveloper314 — APPROVED** [review]

**Review by cn337131 — APPROVED** [review]

**Pull request merged**